smartsim/_core/launcher/dragon/dragonBackend.py

```diff
@@ -155,150 +347,314 @@ def _allocate_step(
-                        group_info.return_codes = [-1 for _ in puids]
-                else:
-                    group_info.return_codes = [0]
-                group_info.status = (
-                    SmartSimStatus.STATUS_FAILED
-                    if any(group_info.return_codes) or grp.status == DRG_ERROR_STATUS
-                    else SmartSimStatus.STATUS_COMPLETED
+                self._queued_steps[step_id] = request
+                self._group_infos[step_id] = ProcessGroupInfo(
+                    status=SmartSimStatus.STATUS_NEVER_STARTED
                 )
-
-            if group_info.status in TERMINAL_STATUSES:
-                terminated.append(step_id)
-
-        if terminated:
-            print(f"{self._updates}: {terminated=}", flush=True)
-        for step_id in terminated:
-            self._running_steps.remove(step_id)
-            self._completed_steps.append(step_id)
-            group_info = self._group_infos[step_id]
-            if group_info is not None:
-                with self._hostlist_lock:
-                    for host in group_info.hosts:
-                        print(f"{self._updates}: Releasing host {host}", flush=True)
-                        self._allocated_hosts.pop(host)
-                        self._free_hosts.append(host)
+            return DragonRunResponse(step_id=step_id, error_message=err)
 
     @process_request.register
     def _(self, request: DragonUpdateStatusRequest) -> DragonUpdateStatusResponse:
-        return DragonUpdateStatusResponse(
-            statuses={
-                step_id: self._group_infos[step_id].smartsim_info
-                for step_id in request.step_ids
-                if step_id in self._group_infos
-            }
-        )
+        with self._queue_lock:
+            return DragonUpdateStatusResponse(
+                statuses={
+                    step_id: self._group_infos[step_id].smartsim_info
+                    for step_id in request.step_ids
+                    if step_id in self._group_infos
+                }
+            )
 
     @process_request.register
     def _(self, request: DragonStopRequest) -> DragonStopResponse:
-        if request.step_id in self._group_infos:
-            # Technically we could just terminate, but what if
-            # the application intercepts that and ignores it?
-            proc_group = self._group_infos[request.step_id].process_group
-            if proc_group is None:
-                self._group_infos[request.step_id].status = SmartSimStatus.STATUS_FAILED
-            elif proc_group.status not in TERMINAL_STATUSES:
-                try:
-                    proc_group.kill()
-                except DragonProcessGroupError:
-                    try:
-                        proc_group.stop()
-                    except DragonProcessGroupError:
-                        print("Process group already stopped")
-
+        with self._queue_lock:
+            self._stop_requests.append(request)
```

Comment:
Do we really need to keep `_stop_requests` as a member variable the entire time a task is running? If we're trying to stop a bunch of running tasks, could we just as easily send a request for each item in `self._running_steps`? That also lets us avoid managing extra things when a task completes (e.g. `self._stop_requests.pop(some_event)`).

Reply:
I see how it's working, now. I didn't realize the `_stop_steps()` method was called during each update loop. Nothing to see here!